### PR TITLE
fix #3283: allow to select typescript version

### LINKED ISSUE #3283: Allow a client to configure typescript.tsdk
First it should be resolved in typescript ls: https://github.com/theia-ide/typescript-language-server/issues/64

More info: https://code.visualstudio.com/docs/languages/typescript#_using-newer-typescript-versions

### PR DESCRIPTION
**Pull request by akosyakov**
Fix #3283

Also, allow providing LS start options.

ts_version

### CODE REVIEW

**Comment by AlexTugarev** [comment]
@akosyakov, I couldn't manage to see the ts version switching. I've installed ts globally with a different version (also restarted the theia app and verified, that the `tsc` version from the env is the other), but it only shows the the same version from `node_modules`. 

screen shot 2018-11-21 at 15 09 30

**Comment by akosyakov** [comment]
@AlexTugarev what do you see if you start a new terminal and run `tsc --version`?

**Comment by AlexTugarev** [comment]
> what do you see if you start a new terminal and run tsc --version?

`Version 3.0.1` which is the other TS version I installed globally. I also terminated the app, checked `tsc --version` on the shell before I restarted.

**Comment by AlexTugarev** [comment]
Another question is if we want to persist the selection in the preferences.

For example in vscode, if I select the workspace ts instead of the bundled, this will be written to `setting.json`. Do we also want to have this behavior? Not sure if this is in the scope of this PR.

**Comment by akosyakov** [comment]
@AlexTugarev thanks for the review. thinking about it:
- if a binary is looked up via `child_process.spawn` from Theia then it is not global, but bundled, because it depends on Theia env
- to look up global we should use `RawProcess`

File: packages/typescript/src/node/typescript-version-service-impl.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    async getVersions(options: TypescriptVersionOptions): Promise<TypescriptVersion[]> {
+        const versions: TypescriptVersion[] = [];
+        await this.resolveGlobalVersion(versions);
+        await this.resolveBundledVersions(versions);
+        await this.resolveWorkspaceVersions(versions, options);
+        return versions;
+    }
+
+    protected async resolveGlobalVersion(versions: TypescriptVersion[]): Promise<TypescriptVersion | undefined> {
+        const command = isWindows ? 'tsserver.cmd' : 'tsserver';
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/typescript/src/node/typescript-version-service-impl.ts]: added tests

File: packages/languages/src/browser/language-client-contribution.ts [hunk 2]
```diff
@@ -97,47 +98,78 @@ export abstract class BaseLanguageClientContribution implements LanguageClientCo
...
+    protected stop = Promise.resolve();
+    protected async doStart(): Promise<Disposable> {
+        const toStop = new DisposableCollection(
+            Disposable.NULL // mark as not disposed
+        );
+        // make sure that the previous client is stopped to avoid duplicate commands and language services
+        await this.stop;
+        const startParameters = await this.getStartParameters();
+        const sessionId = await this.languageContributionService.create(this.id, startParameters);
+        if (toStop.disposed) {
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/languages/src/browser/language-client-contribution.ts]: my changes are bogus, one should be able to stop contribution as fast as possible, i will revert `doStart`

File: packages/typescript/src/node/typescript-version-service-impl.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply packages/typescript/src/node/typescript-version-service-impl.ts]: Awesome!

**Comment by akosyakov** [comment]
@AlexTugarev I've addressed all your comments. If you can please test a bit LS reconnection on flaky connections.

**Comment by AlexTugarev** [comment]
screen shot 2018-11-23 at 15 50 58

I cannot tell the difference between those, do you?
wait, is it because I've started theia and opened its project?

**Comment by AlexTugarev** [comment]
@akosyakov the tsserver search looks only looks at the root-level, right? 

I opened theia from `#/workspace/foo/` and then cloned `typeorm`. after `npm install`, only the bundled version was available. 

after opening `#/workspace/foo/typeorm`, I can select the workspace version. 

Awesome! 🎉

**Comment by akosyakov** [comment]
@AlexTugarev it is turned out that `LanguageClient` does not remove all listeners, so we should not create more than one `LanguageClient` per a contribution, otherwise they are leaking. It seemed to be like that always, i've added a commit to fix it: https://github.com/theia-ide/theia/pull/3538/commits/2cb9573e7d13dd9b77383953ef26e80e28279575

**Comment by simark** [comment]
@akosyakov, commit `[languages] fix leaking language clients ` broke the feature of switching C++ build configuration.

That feature relied on the fact that restarting the language server would instantiate a new language client.  When creating the new language client, [`createOptions`](https://github.com/theia-ide/theia/blob/master/packages/cpp/src/browser/cpp-language-client-contribution.ts#L115) would be called again, and this is where we would feed the new build configuration in the `initialize` params.

After this patch, `createOptions` is no longer called.  It looks like the language client is re-used for the new connection, and it reuses the old options.

What do you suggest to fix this?  Should `createOptions` be called again even if we re-use the same client?

**Pull request merged**